pkg/acquisition/modules/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,402 @@
+			SharedConfigState: session.SharedConfigEnable,
+			Profile:           *cw.Config.AwsProfile,
+		}))
+	} else {
+		sess = session.Must(session.NewSessionWithOptions(session.Options{
+			SharedConfigState: session.SharedConfigEnable,
+		}))
+	}
+
+	if sess == nil {
+		return fmt.Errorf("failed to create aws session")
+	}
+	cw.cwClient = cloudwatchlogs.New(sess)
+	if cw.cwClient == nil {
+		return fmt.Errorf("failed to create cloudwatch client")
+	}
+	monitChan := make(chan LogStreamTailConfig)
+	t.Go(func() error {
+		return cw.LogStreamManager(monitChan, out)
+	})
+	return cw.WatchLogGroupForStreams(monitChan)
+}
+
+func (cw *CloudwatchSource) GetMetrics() []prometheus.Collector {
+	return nil
+}
+
+func (cw *CloudwatchSource) ConfigureByDSN(string, string, *log.Entry) error {
+	return nil
+}
+
+func (cw *CloudwatchSource) GetMode() string {
+	return configuration.TAIL_MODE
+}
+
+func (cw *CloudwatchSource) GetName() string {
+	return "cloudwatch"
+}
+
+func (cw *CloudwatchSource) OneShotAcquisition(chan types.Event, *tomb.Tomb) error {
+	return nil
+}
+
+func (cw *CloudwatchSource) CanRun() error {
+	return nil
+}
+
+func (cw *CloudwatchSource) Dump() interface{} {
+	return cw
+}
+
+func (cw *CloudwatchSource) WatchLogGroupForStreams(out chan LogStreamTailConfig) error {
+	cw.logger.Infof("Starting to watch group")
+	ticker := time.NewTicker(*cw.Config.PollNewStreamInterval)
+	var startFrom *string
+
+	for {
+		select {
+		/*TODO handle tomb*/
+		/*TODO test one specific stream only*/
+		case <-ticker.C:
+			hasMoreStreams := true
+			startFrom = nil
+			for hasMoreStreams {
+				ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+				//there can be a lot of streams in a group, and we're only interested in those recently written to, so we sort by LastEventTime
+				err := cw.cwClient.DescribeLogStreamsPagesWithContext(
+					ctx,
+					&cloudwatchlogs.DescribeLogStreamsInput{
+						LogGroupName:        aws.String(cw.Config.GroupName),
+						Descending:          aws.Bool(true),
+						LogStreamNamePrefix: cw.Config.StreamPrefix,
+						NextToken:           startFrom,
+						OrderBy:             aws.String(cloudwatchlogs.OrderByLastEventTime),
+					},
+					func(page *cloudwatchlogs.DescribeLogStreamsOutput, lastPage bool) bool {
+						for _, event := range page.LogStreams {
+							startFrom = page.NextToken
+							//we check if the stream has been written to recently enough to be monitored
+							if event.LastIngestionTime != nil {
+								//aws uses millisecond since the epoch
+								oldest := time.Now().UTC().Add(-*cw.Config.MaxStreamAge)
+								LastIngestionTime := time.Unix(0, *event.LastIngestionTime*int64(time.Millisecond))
+								if LastIngestionTime.Before(oldest) {
+									cw.logger.Tracef("stop iteration, %s reached oldest age, stop (%s < %s)", *event.LogStreamName, LastIngestionTime, time.Now().Add(-*cw.Config.MaxStreamAge))
+									hasMoreStreams = false
+									return false
```

Comment:
same here (prototype return an error)

Reply:
the prototype is `func(page *cloudwatchlogs.DescribeLogStreamsOutput, lastPage bool) bool` -> we return false if we want to stop iteration.